NPEP: Iron out Cluster Egress Support API Design (#144)
kubernetes-sigs/network-policy-api: npep/npep-126-egress-traffic-control.md

}
```

Note that the API expects `externalNetworks` to be a set of "external" destinations.
However if user puts a podCIDR, nodeCIDR, serviceCIDR or other intra-cluster
networks, it is implementation-defined as to what the behavior will be. Not
all implementations can correctly define the boundary between "internal" and
"external" destinations with respect to a Kubernetes cluster.

Example: <TBD> -> after we reach consensus on externalNetwork versus network

Let's define ANP and BANP that refer to these external networks:
```
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: cluster-egress-controls
spec:
  priority: 70
  subject:
    namespaces: {}
  egress:
  - name: "allow-all-egress-to-intranet"
    action: "Allow"
    to:
    - externalNetworks:
        matchLabels:
          network: intranet
  - name: "allow-egress-to-external-dns"
    action: "Allow"
    to:
    - externalNetworks:
        matchLabels:
          network: dns
    ports:
      - portNumber:
          protocol: UDP
          port: 53
  - name: "pass-egress-to-internet"
    action: "Pass"
    to:
    - externalNetworks:
        matchLabels:

>           network: internet

Reviewer: From above schema, I see `ExternalNetworks` as array of strings. How can it match labels here? Are these predefined labels need to be created by admin and group IPs?

Author: @tamilmani1989 : Were you there at KubeCon NA Chicago and were you the one asking about having a new object?
So this part of the NPEP is still not entirely fixed. See https://github.com/kubernetes-sigs/network-policy-api/pull/144#discussion_r1370456050 for details.

`Example: <TBD> -> after we reach consensus on externalNetwork versus network` -> anything below this is right now under progress. As of now we are thinking of simply having a string of CIDRs in the main API instead of new object and then using selectors to match on them because no one reached out with use cases for needing a new object. Do you have strong use case in which case we can pivot towards this.

Reviewer: nope it's not me at kubecon.

Author: @tamilmani1989 : ack, IIUC so as far as your use case is concerned what you are looking for is a way to explicitly deny a set of cidrs without having to implicitly deny everything. I think that use case will be covered regardless in ANP for sure so that admins can define firewall type allow/deny rules.

Reviewer: @tssurya yes that's correct. Thanks for confirming.

Reviewer 2: @tssurya Looking for that ^^ Updated here, otherwise Things are looking good 👍

Reviewer 3: I have a user story that I think motivates a dedicated CIDR object rather than directly listing the CIDRs in the policies:

```
**As a** cluster administrator **I want** to to ensure that pods can
reach commonly-used databases under my control but outside Kubernetes.
Many but not all applications in my environment rely on these
databases. I want to delegate writing network policy that allows
traffic to the databases to namespace owners.
Example: As a cluster administrator I define a CIDR group that
defines a set of RDS instances that is used across multiple apps.
The owners of namespaceA and namespaceB can then define policies that
allow traffic to this group of RDS instances, and they reference the
instances by CIDR group. As a cluster administrator I can migrate the
database infrastructure and update the CIDR group independently of
the namespace owners. The applications in namespaceC do not use this
infrastructure, so the cluster administrator and the owners
of namespaceC do not need to think about network policy for this
infrastructure.
```

Reviewer 4: Another use case could be a cluster controller that watches external resources (e.g. via cloud API or BGP) and updates the CIDR object with the changes. In this case, the controller only needs enough RBAC rules to update CIDR object and would not need touch the ANP itself

Author: @networkop: oh that's a good point.
@joestringer: I guess cilium gives both ways? cidrs on policy straight and cidrs on new object? or is it only the latter?

Reviewer 5: very interesting feedback! So it sounds like 2 new user stories
1. As a cluster admin, I want to delegate policy decision for a specific CIDR (or maybe even call is resource) that is managed by me (cluster admin) to namespace owners.
One obvious advantage of using ANP compared to referencing admin-managed CIDR is that if at some point admin decides to change ANP to use e.g. DNS name to reference the same resource (e.g. RDS) they can just update ANP and users won't even notice it (can't do that with CIDRs). Another example is if at some point RDS becomes internal to the cluster, e.g. will be run as a k8s service, admin may also change the ANP to use pod selectors instead of CIDRs, and namespace owners don't really need to know about it.
For me it seems like namespace owner have the autonomy by both using admin-provided CIDR and using admin-provided allow ANP, they depend on admin to define some resource either way, but maybe I am missing something here @joestringer? (though I do see how it may _feel_ different for namespace owners, I don't see how it is _really_ different :) )
2. As a cluster admin, I want to define ANP based on a dynamic CIDR that is managed by another controller.
This is also very interesting, but if I got the use case right, it is very different from the previous one (or is there a need to share dynamic CIDRs with namespace owners too?). I feel like more details are really important here @networkop, e.g. who creates the CIRDs resource and who updates it, how is referencing/naming organized? i.e. how can cluster admin understand what a given CIDR managed by another controller means?

Reviewer 6: To ideally realize use-case 1, we'd need a new NetworkPolicy v2 (which is namespace-scoped and understands the CIDRGroup object). In such a scenario we should think hard about whether we actually want "ANP resources" to be directly referenced in a future version of NetworkPolicy v2. My instinct is that would cause unnecessary confusion and interlock, but curious to hear what others think.

Author: I have opened https://github.com/kubernetes-sigs/network-policy-api/issues/182 to capture the use cases discussed here and we can go from there. We can keep the focus of this one to defining inline CIDRs. If the need arises for another object to define peers we can still rethink the inline CIDRs aspect since its still all experimental. We can also see if it makes sense to have both types of peers - however that can cause two different ways of expressing the same thing which probably is ok?

Reviewer 3: > I guess cilium gives both ways? cidrs on policy straight and cidrs on new object? or is it only the latter?

Yes, both. Here's the relevant part of the CRD: https://github.com/cilium/cilium/blob/v1.15.0-pre.3/pkg/k8s/apis/cilium.io/client/crds/v2/ciliumnetworkpolicies.yaml#L1518-L1530

@npinaeva I see the two cases as orthogonal but complimentary. If there is an abstraction for a separate resource to configure the members of the CIDR set, then there is flexibility on whether the user may decide to manually update the set or to deploy a controller that manages the set. While I do think the delegation aspect is a different story, it could equally apply in the case of dynamically updated CIDR set resources.

I would characterize two problems, one of which is common with all network policy (defining policy based on logical groupings) and one that is specific to CIDR handling (set membership for the logical groupings). If we consider standard k8s policy, we have the ability to select labels of k8s resources and have the policy apply to the traffic corresponding to those Pods, Namespaces, etc. The user does not need to determine set membership because k8s resources already provide that information (which labels correspond to which IPs, for instance). In the case of CIDR policy, I think that users want to be able to express a similar logical grouping in the policy to say "Pod X can access RDS", which can be more easily reasoned about, and then delegate the set membership question to another resource. The CIDR group then becomes that other resource, which could be managed manually or ideally something can auto-discover / auto-populate the relevant CIDRs into the group.

We don't debate whether the sets for pod selectors or namespace selectors should be managed by the central administrator because they're k8s primitives. I think there's a case for a cidr selector that operates in a similar way.

> One obvious advantage of using ANP [...] admin decides to change ANP [..] they can just update ANP and users won't even notice it (can't do that with CIDRs).

This is great for organizations where there is a strong central team with the sufficient support, confidence and purview over the network policy to be able to apply it to the cluster. However, I think it's making the decision for the end user organizations about who should write the policy. For organizations that delegate network policy to namespace owners, they may not see this as an option. Having a CIDR selector could enable such organizations to get the same flexibility over policy configuration.

Author: > To ideally realize use-case 1, we'd need a new NetworkPolicy v2 (which is namespace-scoped and understands the CIDRGroup object). 

yes the first one is tied to DNP (https://github.com/kubernetes-sigs/network-policy-api/issues/136). 

> In such a scenario we should think hard about whether we actually want "ANP resources" to be directly referenced in a future version of NetworkPolicy v2. My instinct is that would cause unnecessary confusion and interlock, but curious to hear what others think.

hmm the only counter argument is the pass action where we want the DNPs to take effect for a specific cidr block set and want to use the objects to be referenced from DNP and ANP, however in that case I'd keep the new object very very generic like CIDRs or whatever which is not qualifying as an ANP only resource.

Reviewer 7: > then if any namespace wants to connect to the db, they can apply `allow-admins-db: true` label to their namespace

Namespace admins cannot apply labels to their own namespaces. The Namespace objects themselves are owned by the cluster admins (and this fact is assumed in lots of APIs).

Reviewer 7: > To ideally realize use-case 1, we'd need a new NetworkPolicy v2 (which is namespace-scoped and understands the CIDRGroup object). In such a scenario we should think hard about whether we actually want "ANP resources" to be directly referenced in a future version of NetworkPolicy v2.

The current idea for "NetworkPolicy v2" is that `networkingv1.NetworkPolicy` would be deprecated and replaced by a new-and-improved object called "DeveloperNetworkPolicy" or something like that, and ANP+DNP combined would constitute "NetworkPolicy v2" (with the cluster-admin and namespace-admin/app-deployer roles clearly separated).

Reviewer 7: > Another use case could be a cluster controller that watches external resources (e.g. via cloud API or BGP) and updates the CIDR object with the changes.

This sounds _a lot_ like [FQDN rules](https://github.com/kubernetes-sigs/network-policy-api/issues/133)... the policy basically says "I want to be able to connect to objects in the `api.example.com` CIDR group", and a controller updates that CIDR group to always have the IPs that `api.example.com` resolves to... the only difference is that as currently defined/imagined, the FQDN rules don't actually explicitly list their IPs anywhere (and maybe we wouldn't want them to because they change too rapidly?). But still... it's _very_ similar...

Reviewer 4: > it is very different from the previous one (or is there a need to share dynamic CIDRs with namespace owners too?).

Yeah, it's a different one. There's no need to share or delegate anything to a namespace owner.

> who creates the CIRDs resource and who updates it

We can assume that cluster admin must create the CIDR group object at the same time as the ANP object referencing it. It could either be pre-populated with entries or left empty for external controller to populate 

> how is referencing/naming organized?

This could a CRD object reference (object name, assuming CIDR object cluster-wide ) or, alternatively, a label selector to be able to match multiple CIDR objects

> This sounds a lot like FQDN rules...

Yep, very similar to FQDN rules, with the exception that the IP discovery is done not via DNS but via labels/tags in the cloud or e.g. BGP communities for on-prem/baremetal environments

Reviewer 5: >Namespace admins cannot apply labels to their own namespaces. 

In this case I mean pod labels, it is required to provide the same granularity as NetworkPolicy.

Reviewer 5: > I see the two cases as orthogonal but complimentary.

Sure, they are both about delegating CIDRs management, but in the first case it's namespace owners delegating management to cluster admin, and in the second case it's cluster admin delegating management to a controller, that is managed by cluster admin (same persona). If want to join these use cases with the same API, we should discuss is it ok if all namespaces will be able to see/use all CIRD groups (e.g. use empty selector for CIRD groups)? It means if cluster admin wants to use CIDR group in ANP (like in BGP example), there is no way to prevent namespaces from using it (could it expose some sensitive information?).

When talking about controller-managed and admin-managed CIDR groups, let's see how CIDR group spec could look like (considering "We can assume that cluster admin must create the CIDR group object at the same time as the ANP object referencing it.")
e.g. for only admin-managed CIDR groups we could just have `spec.CIDRs []string`, but if we want controller to manage e.g. DNS names, it should be some `spec.FQDN string` + `status.CIDRs []string`. Then for some other controller populating BGP params it could be `spec.Community string` + `status.CIDRs []string`. So how can we define one spec here?
There are more options around organizing something similar, these questions are based on the mentioned answer from @networkop. Very curious to learn how your BGP example works.

Reflecting on how this is similar to FQDN selector we are working on, should we consider instead of `peer.fqdn: <a.b.com>` using `peer.cirdGroup: matchLabels:{cidrGroup.kubernetes.io/fqdn: <a.b.com>}`, or the opposite for other controller-managed cidrGroups: use `peer.<CIDR group semantics>` without cidrGroup reference, and only leave cidrGroups for manual management? Or maybe something in between? (Rapidly changing ips for FQDN is an interesting limitation for using "controller-managed CIDR group")

Reviewer 4: sorry for the late reply, I must have missed the notification. I can clarify the BGP use case a bit.
The use case I have in mind is the following:
As a cluster admin or a namespace owner, I want to be able to create network policies that match a dynamic set of external IPs (e.g. set of VMs or set of directly reachable Pods in another cluster). I may not be able to use FQDN rules for that due to TTL being too long or simply lack of DNS service discovery in an external system. 

As a cluster admin, I would create CIDR group resource and a BGP controller that would manage it. The mapping between BGP communities and CIDR group resource names is a BGP controller configuration (e.g. annotation on the CIDR group resource). The speed of IPs churn is bounded by the BGP advertisement interval and can be further reduced by the BGP controller implementation.

Apologies, I'm not an expert in k8s APIs but I've tried to mock up how would a BGP-managed CIDR group object look like:

```yaml
apiVersion: policy.networking.k8s.io/v1alpha1
kind: CIDRGroup
metadata:
  name: cluster-wide-cidr-cloud-1
  labels:
    origin: cloud-1
  annotations:
    "bgp.cidrmanager.k8s.io/is-managed": "true"
    "bgp.cidrmanager.k8s.io/32bit-community": "2147483647"
spec:
  cidrs:
  - 192.0.2.0/24
  - 203.0.113.0/24
  - 198.51.100.0/24
status:
  conditions:
  - lastTransitionTime: "2022-12-29T14:53:50Z"
    status: "True"
    type: Reconciled

---
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: cluster-wide-allow-example
spec:
  priority: 30
  subject:
    namespaces: {}
  ingress:
    - action: Allow
      from:
      - cidrGroups:
            matchLabels:
              origin: cloud-1
```

Author: @networkop : thank you so much for the details. We will cover the relevant use cases as part of https://github.com/kubernetes-sigs/network-policy-api/pull/183

## API

Proof of Concept for the API design details can be found here: https://github.com/kubernetes-sigs/network-policy-api/pull/143

### Implementing egress traffic control towards cluster nodes

This NPEP proposes to add a new type of `AdminNetworkPolicyEgressPeer` called `Nodes`
to be able to explicitly select nodes (based on the node's labels) in the cluster.
This ensures that if the list of IPs on a node OR list of nodes change, the users
don't need to manually intervene to include those new IPs. The label selectors will

> take care of this automatically.

Author: ah hmm good point, in OVNK we actually match on all IPs in .Status.Addresses and all secondary network IPs as well of the node that is not known by kubernetes. But I think I can deff call out that the API expects at a minimum for implementations to match on https://kubernetes.io/docs/reference/node/node-status/#addresses -> so both IPs and hostnames as peers? and then individual implementations can handle their own versions...

> take care of this automatically.

Author: done.

```
// AdminNetworkPolicyEgressPeer defines a peer to allow traffic to.
// Exactly one of the selector pointers must be set for a given peer. If a
// consumer observes none of its fields are set, they must assume an unknown
// option has been specified and fail closed.
// +kubebuilder:validation:MaxProperties=1
// +kubebuilder:validation:MinProperties=1
type AdminNetworkPolicyEgressPeer struct {
    <snipped>
	// Nodes defines a way to select a set of nodes in
	// in the cluster. This field follows standard label selector
	// semantics; if present but empty, it selects all Nodes.
	//
	// Support: Core
	//
	// +optional
	Nodes *metav1.LabelSelector `json:"nodes,omitempty"`
}
```

> Note that `AdminNetworkPolicyPeer` will changed to

Author: oops, my bad.
done!

`AdminNetworkPolicyEgressPeer` and `AdminNetworkPolicyIngressPeer` since ingress and
egress peers have started to diverge at this point and it is easy to
maintain it with two sets of peer definitions.
This ensures nodes can be referred to only as "egress peers".

Example: Admin wants to deny egress traffic from tenants who don't have
`restricted`, `confidential` or `internal` level security clearance
to control-plane nodes at 443 and 6443 ports in the cluster

```
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: node-as-egress-peers
spec:
  priority: 55
  subject:
    namespaces:
      matchExpressions:
        - {key: security, operator: notIn, values: [restricted, confidential, internal]}
  egress:
  - name: "deny-all-egress-to-kapi-server"
    action: "Deny"
    to:
    - nodes:
        matchLabels:

>           node-role.kubernetes.io/control-plane: true

Author: ack changing this to `node-role.kubernetes.io/control-plane:`

          port: 443
      - portNumber:
          protocol: TCP
          port: 6443
```

### Implementing egress traffic control towards external destinations

This NPEP proposes to add a new type of `AdminNetworkPolicyEgressPeer` called `ExternalNetworks`
to be able to explicitly select external destinations (based on the `externalNetworkSet`'s
labels) in the cluster. This peer type will not be supported in `AdminNetworkPolicyIngressPeer`.

```
// AdminNetworkPolicyEgressPeer defines a peer to allow traffic to.
// Exactly one of the selector pointers must be set for a given peer. If a
// consumer observes none of its fields are set, they must assume an unknown
// option has been specified and fail closed.
// +kubebuilder:validation:MaxProperties=1
// +kubebuilder:validation:MinProperties=1
type AdminNetworkPolicyEgressPeer struct {
    <snipped>
	// ExternalNetworks defines a way to select ExternalNetworkSets
	// that consist of network CIDRs that live outside the cluster as a peer.
	// This field follows standard label selector semantics; if present
	// but empty, it selects all ExternalNetworkSets defined in the cluster.
	//
	// Support: Core
	//
	// +optional

> 	ExternalNetworks *metav1.LabelSelector `json:"externalNetworks,omitempty"`

Reviewer 7: I missed the discussion of this in the WG meeting, but I suspect there is not actually a strong use case for having a label selector here rather than just referring to a specific ExternalNetworkSet by name.

Author: correct no strong use cases afaik, its only an extra additional thing if like in my example below they want to group their networks into different networksets. I am ok to directly use metadata name here if that is desired.

}
```

> An `externalNetworkSet` is a new object used to define a set of networks outside

Reviewer 5: Let's try to list pros and cons for having a separate object, here is what I can think of:
pros:
- sharing the same ExternalNetworkSets between ANPs, that maybe useful when defining ANPs for different subjects

cons:
- complicated management: additional watcher, integrity violation handlers, cleanup of unreferenced objects, etc
- may worsen readability: to see what a given ANP does you need to query another object

Please add more :)

from the FQDN side, we will probably need another object that would only be used for egress (assuming CIDRs may be used for ingress rules too, but without FQDNs)

Author: > 
> from the FQDN side, we will probably need another object that would only be used for egress (assuming CIDRs may be used for ingress rules too, but without FQDNs)

hmm so that is the thing I am looking for feedback from @rahulkjoshi about. I was thinking an "externalDestination" can be anything, the object `ExternalNetworkSet` currently has a `networks` field which is a list of CIDRs, however if we want to accommodate `fqdn's` as well into this object and then put a at a time only a single peer can be allowed, I am ok to do that and rename this object accordingly.

Reviewer 6: One thing that concerns me about this is that I know there's some discussion about "Network" as a Kubernetes concept (with Pods have multiple NICs and multiple Networks). I worry that have an object called "ExternalNetwork" would be too confusing.

That's exacerbated by making `EgressPeer.ExternalNetwork` be a label-selector. Then the API doesn't automatically indicate what gets selected.

+1 to the cons Nadia suggested. I'm not sure they outweigh the re-use situation, but happy to discuss.

Author: I don't think the cons outweigh the re-use situation but I also don't think we want to do "External" anymore so it might just be `cidrs` - still discussing on the details.
```
That's exacerbated by making EgressPeer.ExternalNetwork be a label-selector. Then the API doesn't automatically indicate what gets selected.
````
yes there was a suggestion to using metadata name as means to select.

Having said that there is really no strong push for doing a new object neither was there specific pushback, so all in all we can keep it simple for now and fold it into the main API. If we want more creative and expressive ways before we move to v1 or beta we can reevaluate this.

the cluster. This is defined as a new object so that its easy for users to group
their external entities and then refer to that group from the different egress rules.
This ensures consistency without the users having to manually change the ANP/BANP

> to accommodate the changing needs of the network cidrs. They can directly edit the

Reviewer 7: Unless you assume admins will have multiple separate ANP objects referring to the same CIDRs, it doesn't seem like updating the ExternalNetworkSet is any easier than updating the ANP...

(Though I guess even without DNP, there's the policy of needing to refer to it from both ANP and BANP...)

Author: yeah I was thinking if admins have multiple egress rules referring to the same set of cidrs, whether those rules are across the same ANP or multiple ANPs its easier to update the ENS...

> An `externalNetworkSet` is a new object used to define a set of networks outside
> the cluster. This is defined as a new object so that its easy for users to group
> their external entities and then refer to that group from the different egress rules.
> This ensures consistency without the users having to manually change the ANP/BANP
> to accommodate the changing needs of the network cidrs. They can directly edit the
> new CRD which guarantees all rules effecting that peer will take effect automatically.
> In the future when we define `DeveloperNetworkPolicies`, this will come even more handy
> for uniformity specially across `Pass` actions.

Reviewer 7: ```suggestion
An `ExternalNetworkSet` is a new object used to define a set of networks outside
the cluster. This is defined as a new object so that it's easy for users to group
their external entities and then refer to that group from the different egress rules.
This ensures consistency without the users having to manually change the ANP/BANP
to accommodate the changing needs of the network CIDRs. They can directly edit the
new object which guarantees all rules affecting that peer will take effect automatically.
In the future if we define `DeveloperNetworkPolicy`, this will come even more handy
for uniformity, especially across `Pass` actions.
```

```
// ExternalNetworkSet is a cluster level resource that is used to define
// a set of networks outsides the cluster which can be referred to from
// the AdminNetworkPolicy && BaselineAdminNetworkPolicy APIs as an external peer
type ExternalNetworkSet struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata"`

	// Specification of the desired behavior of ExternalNetworkSet.
	Spec ExternalNetworkSetSpec `json:"spec"`
}

// ExternalNetworkSetSpec defines the desired state of ExternalNetworkSet.
type ExternalNetworkSetSpec struct {

> 	// Networks is the list of NetworkCIDR (both v4 & v6) that can be used to define

Reviewer 7: ```suggestion
	// Networks is the list of CIDRs (both v4 & v6) that can be used to define
```

> 	// external destinations.

Reviewer 7: Somewhere you need to talk about what happens if someone puts internal CIDRs into an ExternalNetworkSet. (Keeping in mind that in some cases the ANP implementation will not be able to reliably distinguish internal from external CIDRs, so the answer can't just be "they will be ignored".)

Author: yeah good point, I think the answer will be it will take effect :D so be careful what you ask for ?

Author: or errr we should go with "implementation-defined?" :D

	// A total of 100 CIDRs will be allowed in each NetworkSet instance.
	// ANP & BANP APIs may use the .spec.in(e)gress.from(to).externalNetworks selector

> 	// to select a set of external networks

Reviewer 7: you said it was egress-only above

Author: oh oops yes, earlier I had a CEL expression and I changed this to splitting the ingress and egress as two different structs, but this got left over. Editing this to be egress only. nice catch.

	// +optional
	// +kubebuilder:validation:MaxItems=100

> 	Networks []string `json:"networks,omitempty" validate:"omitempty,dive,cidr"`

Author: good question, I shamelessly copied what calico does for validating the CIDR ranges: https://github.com/projectcalico/calico/blob/cbdf9cfb2b212948b03418b6b20e1f64327a9d07/api/pkg/apis/projectcalico/v3/globalnetworkset.go#L51 because it means we don't need to do the complicated CEL stuff: https://github.com/openshift/api/pull/1410/files#diff-2e98c189fdf696589c6d28124557b9cf96a329e82f790b4e61209351cc68648dR552 but perhaps that is not the right approach and we want to do the fancy CEL thing for v4 and v6 which means combining them into the same string might be a bit hard..

}
```

Example: Let's define some sample sets of externalNetworks.

```
apiVersion: policy.networking.k8s.io/v1alpha1
kind: ExternalNetworkSet
metadata:
  name: api-vm-network
  labels:
    network: intranet
spec:
  networks:
    - 10.0.0.10/32
    - 10.0.0.11/32
    - 10.0.0.12/32
---
apiVersion: policy.networking.k8s.io/v1alpha1
kind: ExternalNetworkSet
metadata:
  name: storage-network
  labels:
    network: intranet
spec:
  networks:
    - 95.90.241.22/28
    - 200.0.65.0/24
---
apiVersion: policy.networking.k8s.io/v1alpha1
kind: ExternalNetworkSet
metadata:
  name: dns-network
  labels:
    network: dns
spec:
  networks:
    - 195.90.241.22/28
    - 111.0.65.0/29
---
apiVersion: policy.networking.k8s.io/v1alpha1
kind: ExternalNetworkSet
metadata:
  name: global-network
  labels:
    network: internet
spec:
  networks:
    - 0.0.0.0/0
```
```

> $ oc get ens

Author: true, removing this shorthand...

This NPEP proposes to add a new type of `AdminNetworkPolicyPeer` called `Nodes`
to be able to explicitly select nodes (based on the node's labels) in the cluster.
This ensures that if the list of IPs on a node OR list of nodes change, the users
don't need to manually intervene to include those new IPs. The label selectors will
take care of this automatically.

```
// AdminNetworkPolicyPeer defines an in-cluster peer to allow traffic to/from.
// Exactly one of the selector pointers must be set for a given peer. If a
// consumer observes none of its fields are set, they must assume an unknown
// option has been specified and fail closed.
// +kubebuilder:validation:MaxProperties=1
// +kubebuilder:validation:MinProperties=1
type AdminNetworkPolicyPeer struct {
    <snipped>
	// Nodes defines a way to select a set of nodes in
	// in the cluster. This field follows standard label selector
	// semantics; if present but empty, it selects all Nodes.
	//
	// Support: Core
	//
	// +optional
	Nodes *metav1.LabelSelector `json:"nodes,omitempty"`
}
```

In order to ensure this type of peer cannot be set from the ingress
rules, we will add the following validation rule:
```

> +kubebuilder:validation:XValidation:rule="self.all(value, !has(value.nodes))",message="cluster-ingress traffic controls are unsupported"

Author: change this to splitting egress and ingress peer types and remove CELs

Author: done!

kind: AdminNetworkPolicy
metadata:
  name: cluster-egress-controls
spec:
  priority: 70
  subject:
    namespaces: {}
  egress:
  - name: "allow-all-egress-to-intranet"
    action: "Allow"
    to:
    - externalNetworks:
        matchLabels:
          network: intranet
  - name: "allow-egress-to-external-dns"
    action: "Allow"
    to:
    - externalNetworks:
        matchLabels:
          network: dns
    ports:
      - portNumber:
          protocol: UDP
          port: 53
  - name: "pass-egress-to-internet"
    action: "Pass"
    to:
    - externalNetworks:
        matchLabels:
          network: internet
---
apiVersion: policy.networking.k8s.io/v1alpha1
kind: BaselineAdminNetworkPolicy
metadata:
  name: default
spec:
  subject:
    namespaces: {}
  egress:
  - name: "deny-egress-to-internet"
    action: "Deny"
    to:
    - externalNetworks:
        matchLabels:
          network: internet
```
This allows admins to specify all cluster workloads can talk to
the company's internet CIDR's and the external DNS network. It
also allows them to put up default guardrails of not allowing
any other egress traffic towards internet in a BANP.

## Alternatives

> N/A

Author: add expressing cidrs on the main API as an alternative.

Author: done!
